registers: MXCSR

jarkkojs opened the pull request:
Add MxCsr register derived from MxCsr type in enarx/xsave.

Provide two unit tests:
- mxcsr_default(): Check that the constant matches the value read from
  the CPU when the process has not yet written anything else to it.
- mxcsr_write(): Check that the write operation does not corrupt values.

Closes: https://github.com/enarx/enarx/issues/1000
Signed-off-by: Jarkko Sakkinen <jarkko.sakkinen@iki.fi>

jarkkojs commented:
I'll fix the reported issues!

Freax13 commented:
It seems like there is a bug in the `const-default` dependency you added: https://github.com/rust-osdev/x86_64/runs/4901262388?check_suite_focus=true

jarkkojs commented:
> It seems like there is a bug in the `const-default` dependency you added: https://github.com/rust-osdev/x86_64/runs/4901262388?check_suite_focus=true

Hi, would it be more sustainable if I dropped that from this PR? It's not super critical as it is just integer that is sealed to be 0x1F80 (documented in Intel SDM).

jarkkojs commented:
> > It seems like there is a bug in the `const-default` dependency you added: https://github.com/rust-osdev/x86_64/runs/4901262388?check_suite_focus=true
> 
> Hi, would it be more sustainable if I dropped that from this PR? It's not super critical as it is just integer that is sealed to be 0x1F80 (documented in Intel SDM).

I changed it as:

```
impl Default for MxCsr {
    /// The default MXCSR value at reset, as documented in Intel SDM volume 2A.
    const DEFAULT:  Self = MxCsr::from_bits_truncate(0x1F80);

    #[inline]
    fn default() -> Self {
        Self::DEFAULT
    }
}
```

Obviously this means that `#derive(ConstDefault)` is not possible but it is not huge deal, i.e. you can implement ConstDefault whatever uses mxcsr.

jarkkojs commented:
@npmccallum, @haraldh: This essentially the same type as in xsave except
```
        /// Toward negative infinity
        const ROUNDING_CONTROL_NEGATIVE = 1 << 13;
        /// Toward positive infinity
        ///
        /// Set the both control registers for "toward zero".
        const ROUNDING_CONTROL_POSITIVE = 1 << 14;
```

In the original type these were `ROUNDING_CONTROL0` and `ROUNDING_CONTROL1`. I cannot remember on top of my head the exact SDM terms (might have been even the original) but these self-document the use and purpose.

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,107 @@
+    ///
+    /// # Safety
+    ///
+    /// Unsafe because undefined becavior can occur given that input is not
+    /// validate before it is passed to ldmxcsr.
```
How can this cause undefined behavior (eg which flags)?

jarkkojs replied:
The upper 16-bit word contains reserved bits (cleared on power on), and setting them causes #GP. Should pre-validate or just document?

Freax13 replied:
We don't consider triggering CPU exceptions to be unsafe (see https://github.com/rust-osdev/x86_64/issues/202#issuecomment-719923562). Are there any other ways this could cause UB?

jarkkojs replied:
Nothing else comes to mind, e.g. according to the SDM:

```
Flags Affected
None
```

jarkkojs replied:
Right, setting upper flags would be _well-defined_ behavior because you consistently SEGFAULT :-) I got it.

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,107 @@
+
+    #[cfg(test)]
+    mod test {
+        use crate::registers::mxcsr::MxCsr;
+        use crate::registers::mxcsr::read;
+        use crate::registers::mxcsr::write;
```
Small nit: These can probably collapsed into one import.

jarkkojs replied:
OK, cool, just did not know the preferred convention, will do :-)

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,107 @@
+    /// Return the default MXCSR value at reset, as documented in Intel SDM volume 2A.
+    #[inline]
+    fn default() -> Self {
+        MxCsr::from_bits_truncate(0x1F80)
```
This values should probably be generated by combining the individual flags with the `|` operator. It might still be useful to mention that constant in a comment somewhere though.

jarkkojs replied:
I can do it that way. Yeah, it is still useful to have also `0x1F80` somewhere. E.g. I sometimes use some constant values that I remember as search keys for the SDM :-D I'll change this as the disjunction of the flags and mention the constant in a comment.

Freax13 commented:
> > It seems like there is a bug in the `const-default` dependency you added: https://github.com/rust-osdev/x86_64/runs/4901262388?check_suite_focus=true
> 
> Hi, would it be more sustainable if I dropped that from this PR? It's not super critical as it is just integer that is sealed to be 0x1F80 (documented in Intel SDM).

Probably, yes. Perhaps we could add a `DEFAULT` variant in `MxCsr`'s definition.

jarkkojs commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,107 @@
+    /// MXCSR register.
+    #[repr(transparent)]
+    pub struct MxCsr: u32 {
+        /// Invalid operation flag
+        const INVALID_OPERATION = 1 << 0;
+        /// Denormal flag
+        const DENORMAL = 1 << 1;
+        /// Divide-by-zero flag
+        const DIVIDE_BY_ZERO = 1 << 2;
+        /// Overflow flag
+        const OVERFLOW = 1 << 3;
+        /// Underflow flag
+        const UNDERFLOW = 1 << 4;
+        /// Precision flag
+        const PRECISION = 1 << 5;
```
OK

jarkkojs commented:
OK, I think I resolved the existing review comments (but obviously can refine more if required).

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,120 @@
+        }
+        #[cfg(not(feature = "inline_asm"))]
+        unsafe {
+            mxcsr = crate::asm::x86_64_asm_read_rflags();
```
That's not reading `MxCsr`, that's reading `RFlags`. You'll need to edit `src/asm/mod.rs` and `src/asm/asm.s` to add the function.

Freax13 replied:
No need to apologize :) Most of the functions follow the same pattern, not a whole lot of assembly knowledge required, so it should be easy to just replicate that here. Let me now if you have any questions.

jarkkojs replied:
Uh oh! LOL. I can assure that rflags part was not intentional! How embarrassing :-/

jarkkojs commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,120 @@
+        const ROUNDING_CONTROL_NEGATIVE = 1 << 13;
+        /// Toward positive infinity
+        ///
+        /// Set the both control registers for "toward zero".
```
> 

I had that but did not stage that change. Coming up.

jarkkojs replied:
OK I uploaded new version. My biggest struggle with QA is that I have no idea how to enforce cargo to compile "not inline_asm" flow instead of "inlin_asm" flow.

Freax13 replied:
The CI job uses `cargo build --no-default-features --features external_asm,instructions`.

Freax13 commented on src/asm/asm.s:
```diff
@@ -334,3 +334,15 @@ _x86_64_asm_xsetbv:
+.global _x86_64_asm_write_mxcsr
+.p2align 4
+_x86_64_asm_write_mxcsr:
+    ldmxcsr %rdi
```
In hindsight this wasn't as straight forward as I thought, `ldmxcsr` expects a memory operand. `%rdi` should be put into parentheses and the corresponding external function  declaration in rust should take a pointer to `MxCsr`. 

Note that this is also what the rust code using inline assembly does: it passes a reference.

We do something similar when executing `lgdt`: https://github.com/rust-osdev/x86_64/blob/08b520ef770e55538e65db11d46932305089b71a/src/asm/asm.s#L117 And
https://github.com/rust-osdev/x86_64/blob/08b520ef770e55538e65db11d46932305089b71a/src/asm/mod.rs#L163

Freax13 commented on src/asm/asm.s:
```diff
@@ -334,3 +334,15 @@ _x86_64_asm_xsetbv:
+.global _x86_64_asm_read_mxcsr
+.p2align 4
+_x86_64_asm_read_mxcsr:
+    stmxcsr %rax
```
The same changes regarding memory should be made here and we also need to use `%rdi` instead of `%rax`. `%rax` is the return value, however `stmxcsr` doesn't move the flags into a register that so that we can use it as the return value, it writes them to a memory location, which in our case is given by `%rdi` - the first function parameter. 

We do something similar when executing `sgdt`: https://github.com/rust-osdev/x86_64/blob/08b520ef770e55538e65db11d46932305089b71a/src/asm/asm.s#L129 And
https://github.com/rust-osdev/x86_64/blob/08b520ef770e55538e65db11d46932305089b71a/src/asm/mod.rs#L175

jarkkojs replied:
Yeah, the key problem was this: I did not have any way to exercise both code paths. Now I've tested both of them by writing a small script that temporarily modifies the source code so that it compiles with external asm. See the script below.

npmccallum commented:
@Freax13 Is the `inline_asm` feature actually needed here now that a subset of inline assembly is now stable? It seems reasonable to keep the feature for any `asm` uses that aren't actually stabilized yet. But for those that are already usable on stable, can't we just implement them without a feature gate?

Freax13 commented:
> Is the `inline_asm` feature actually needed here now that a subset of inline assembly is now stable? 

Inline assembly hasn't landed on stable yet, it will be part of Rust 1.59 released at the end of next month. However once that happens, I agree that we could probably only use inline assembly in some (if not all?) cases.

jarkkojs commented:
I sorted my problem out by writing a small script:

```
find -type f -print0 | xargs -0 sed -i 's/inline_asm/xxx_inline_asm/'
find -type f -print0 | xargs -0 sed -i 's/not(xxx_inline_asm)/external_asm/'
git checkout build.rs
git checkout src/lib.rs
cargo build --features external_asm
cargo test --features external_asm
git reset --hard HEAD
```

Now both code paths work!

jarkkojs commented:
You can make amazingly stupid things if you are coding blind, i.e. no ways neither to compile nor run any tests on code you're working on :-)

jarkkojs commented:
I proposed a solution: https://github.com/rust-osdev/x86_64/issues/337

I.e. you could then override x86_64_inline_asm even if its default value would be implicitly set based on inline_asm.

jarkkojs commented:
I got the build commands together:

```
cargo build --no-default-features --features external_asm,instructions
cargo test --no-default-features --features external_asm,instructions -- --test-threads=1
```

With this I was able to run the tests sequentially through with external asm. 

/Jarkko

Freax13 commented:
> I got the build commands together:
> 
> ```
> cargo build --no-default-features --features external_asm,instructions
> cargo test --no-default-features --features external_asm,instructions -- --test-threads=1
> ```
> 
> With this I was able to run the tests sequentially through with external asm.

This can be further simplified to just:
```bash
cargo test --no-default-features --features external_asm,instructions
```

`cargo test` also automatically builds the code before running the tests and I don't think there's reason to limit the threads for running the tests.

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,120 @@
+#[cfg(feature = "instructions")]
+mod x86_64 {
+    use super::*;
+    use core::arch::asm;
```
```suggestion
    #[cfg(feature = "inline_asm")]
    use core::arch::asm;
```

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,120 @@
+    /// Read the value of MXCSR.
+    #[inline]
+    pub fn read() -> MxCsr {
+        let mut mxcsr: u32 = 0;
```
rustc complains here because the initial value is never used here when compiling with `external_asm`.

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,120 @@
+    pub fn write(mxcsr: MxCsr) {
+        #[cfg(feature = "inline_asm")]
+        unsafe {
+           asm!("ldmxcsr [{}]", in(reg) &mxcsr, options(nostack));
```
This should also contain the `readonly` option.

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,120 @@
+
+        #[cfg(feature = "inline_asm")]
+        unsafe {
+            asm!("stmxcsr [{}]", in(reg) &mut mxcsr, options(nostack));
```
This should also contain the `preserves_flags` option.

Freax13 commented on README.md:
````diff
@@ -15,3 +15,13 @@ Support for x86_64 specific instructions (e.g. TLB flush), registers (e.g. contr
 This needs to have the [compile-time requirements](https://github.com/alexcrichton/cc-rs#compile-time-requirements) of the `cc` crate installed on your system.
 It was currently only tested on Linux and MacOS.
+
+## Enforcing external_asm build
+
+In addition to the standard cargo build the `external_asm` build path needs to
+be exercised:
+
+```
+cargo build --no-default-features --features external_asm,instructions
+cargo test --no-default-features --features external_asm,instructions -- --test-threads=1
+```
````
The README is only targeted at the crate's users not the developers, so perhaps this could be moved to a CONTRIBUTING file. We could also mention that this is only necessary when adding/modifying instruction functions.

jarkkojs replied:
I can do that.

Freax13 reviewed (changes requested):
A few more small changes required, we should be able to merge this soon :-). 

The CI job that checks formatting is also still failing, you can run `cargo fmt` to fix this.

jarkkojs commented:
Thanks, I'll fix these later on today.

jarkkojs commented:
OK, let's see :-)

Freax13 commented on src/registers/mxcsr.rs:
```diff
@@ -0,0 +1,119 @@
+    /// Read the value of MXCSR.
+    #[inline]
+    pub fn read() -> MxCsr {
+        #[cfg(feature = "inline_asm")]
+        unsafe {
+            let mut mxcsr: u32 = 0;
+            asm!("stmxcsr [{}]", in(reg) &mut mxcsr, options(nostack, preserves_flags));
+            MxCsr::from_bits_truncate(mxcsr)
+        }
```
```suggestion
        #[cfg(feature = "inline_asm")]
        {
            let mut mxcsr: u32 = 0;
            unsafe {
                asm!("stmxcsr [{}]", in(reg) &mut mxcsr, options(nostack, preserves_flags));
            }
            MxCsr::from_bits_truncate(mxcsr)
        }
```

Freax13 reviewed (changes requested):
One last small suggestion, then this should be good to go  :tada:

Freax13 commented:
> One last small suggestion, then this should be good to go tada

Oh and you need to run `cargo fmt`.

jarkkojs commented:
> > One last small suggestion, then this should be good to go tada
> 
> Oh and you need to run `cargo fmt`.

I'll keep this in mind, thank you.

jarkkojs commented:
So... I guess it is now about right?

Freax13 commented:
> So... I guess it is now about right?

Yes. Thank you for your contribution!
